### DISCUSSION TITLE
How to stop and restart the script?

### DISCUSSION THREAD

**Discussion by synapticonashaikh** [micropython/micropython#12044] [General]
Hello, I'm working on a project where I modified the minimal port to be able to use it for my processor (XMOS). In my application, I run a pre-compiled script file compiled using the mpy-cross cross compiler. Based on some commands. I want to start, stop and restart the script. Right now, I'm just able to start. I tried a few methods to stop, and it worked (not sure if it was the correct method), but then I was unable to restart the script again. Mainly, once I call mp_deinit( ), the code gets stuck if I call mp_init again. 

My processor has multiple logical cores, and each one can run one thread function on it. So, I run the script on one thread and try to start, stop, and restart script execution from another thread function. Please check the semi-pseudo code, which is attached below and let me know if there is any better way to do it. 

```
#define none -1
#define start 1
#define stop 2
#define restart 3 

extern int8_t command;
extern int ScriptLen;
extern uint8_t* Script;
uint8_t ScriptStartFlag, GCInitFlag;
static char *stack_top;

/*************************************************/
void Thread_one(void) 
{
	while (1) { //assuming that the command variable is being set outside
		if(command == start)  ScriptStartFlag = 1;
		else if(command == stop)  FnStopTheScript( );
		else if(command == restart) FnRestartTheScript( );
	
		if (command != none) command = none; // reset the command flag
	}
}

/*************************************************/
void Thread_two(void)
{
	while (1) {
		if ( ScriptStartFlag == 1 ) {
			ScriptStartFlag = 0;
	 		FnStartTheScript( );
		}
	}
}

/*************************************************/
void FnStartTheScript(void) // working
{
    int stack_dummy;
    stack_top = (char *)&stack_dummy;    
    #if MICROPY_ENABLE_GC
    if (GCInitFlag == 0) {   
    	GCInitFlag = 1;
        gc_init(heap, heap + sizeof(heap));
    }
    mp_init( );

    mp_reader_t reader;
    mp_reader_new_mem(&reader,(byte *)Script, ScriptLen, 0);
    mp_module_context_t *context = m_new_obj(mp_module_context_t);
    context->module.globals = mp_globals_get();
    mp_compiled_module_t compiled_module;
    compiled_module.context = context;
    mp_raw_code_load(&reader, &compiled_module);
    mp_obj_t module_fun = mp_make_function_from_raw_code(compiled_module.rc, context, NULL);
    mp_call_function_0(module_fun);
 
    mp_deinit( );
}

/*************************************************/
void FnStopTheScript(void) //working
{
  mp_sched_lock( ); gc_lock( );
  mp_raise_type(&mp_type_SystemExit);
  mp_sched_unlock(); gc_unlock( );
  /*deinit*/ 
  mp_deinit( );
}

/*************************************************/
void FnRestartTheScript(void) // not working
{
	FnStopTheScript( );
	FnStartTheScript( ); // it gets stuck at mp_init( ); 
}
```

**Comment by bidrohini** [comment]
You can try the code:
#include <stdbool.h>

#define NONE -1
#define START 1
#define STOP 2
#define RESTART 3

extern int8_t command;
extern int ScriptLen;
extern uint8_t* Script;
bool scriptRunning = false;
bool scriptInitialized = false;
static char *stack_top;

/*************************************************/
void Thread_one(void) 
{
	while (1) {
		if (command == START) {
			if (!scriptRunning) {
				scriptRunning = true;
				if (!scriptInitialized) {
					scriptInitialized = true;
					FnStartTheScript();
				}
			}
		} else if (command == STOP) {
			if (scriptRunning) {
				scriptRunning = false;
				FnStopTheScript();
			}
		} else if (command == RESTART) {
			if (scriptRunning) {
				scriptRunning = false;
				FnStopTheScript();
			}
			scriptInitialized = false;
			command = START; // Next iteration will start the script
		}
		command = NONE; // reset the command flag
	}
}

/*************************************************/
void Thread_two(void)
{
	while (1) {
		if (scriptRunning) {
			FnExecuteTheScript();
		}
	}
}

/*************************************************/
void FnStartTheScript(void)
{
	int stack_dummy;
	stack_top = (char *)&stack_dummy;

	// Perform any one-time initialization here, e.g., GC initialization
	// ...

	// Load and compile the script once
	mp_reader_t reader;
	mp_reader_new_mem(&reader, (byte *)Script, ScriptLen, 0);
	mp_module_context_t *context = m_new_obj(mp_module_context_t);
	context->module.globals = mp_globals_get();
	mp_compiled_module_t compiled_module;
	compiled_module.context = context;
	mp_raw_code_load(&reader, &compiled_module);
	mp_obj_t module_fun = mp_make_function_from_raw_code(compiled_module.rc, context, NULL);

	// The actual script execution loop should be separate from initialization
	// Store the module_fun in a global variable or some other accessible place

	// Note: mp_init() should be called only once during the application's startup
}

/*************************************************/
void FnExecuteTheScript(void)
{
	// Retrieve the stored module_fun and execute it in a loop
	// ...

	// Check for script termination conditions and break the loop
	// ...

	// If the script has terminated, set scriptRunning to false
}

/*************************************************/
void FnStopTheScript(void)
{
	// Terminate the script execution gracefully
	// ...

	// Deallocate any resources or memory used by the script
	// ...
}

/*************************************************/
void FnRestartTheScript(void)
{
	FnStopTheScript();
	scriptInitialized = false;
	command = START; // Next iteration will start the script
}

**Reply by synapticonashaikh** [reply]
I'm mainly after how to do the function FnStopTheScript, where I want to know how to deallocate memory and de-init everything so it can be re-init again. The above code is just a pseudo example code for better understanding. After getting some good hints, I'll copy the start, stop and restart functions.
